### ISSUE TITLE
Make a clear plan to support Transformer on Fluid.

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#6876]
We are going to support popular NMT models on Fluid, including but not limited to RNN search, ConvS2S, and Transformer.

I think the first important thing for us is to understand and figure out the problems.

We choose Google' Transformer as our starting point. Here I list some questions **should be answered**:

A tensor2tensor implementation: https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py

- About the Transformer:
    - motivation: What is the problem proposed to solve.
    - how the problem is solved.
    - model architecture:
        - Are there any challenges or functional requirements for a DL framework to implement the transformer architecture?
        - Are there any tricks or key points require our framework to support a function we have not considered yet?
        - Are there any operators we have not implemented yet?

- About Fluid (This part is relatively open currently)
    - How encoder-decoder architecture is implemented in Fluid, run it and follow its execution/codes, do you have any question and problem about it (it is ok just from the perspective of a user)
    - RNN and dynamic RNN. The transformer does not need any recurrent layer and CNN layer. But how Fluid process sequence is still crucial.
    - propose your questions about Fluid, we will collect them.

---

At the end of this step, we will share our notes with everyone, both about Transformer and Fluid,  we can try to make it part of the document.  all of us should:

1.  have an overview picture of the architecture of Transformer; (from top to bottom, not the reverse.)
1. carry out a clear plan about how we are going to implement Transformer:
    - any special requirements for Fluid?
    - what mechanisms of fluid are we going to make use of? for example: while loop, dynamic RNN, or any other things?
1. a checklist of operators needed: The list can be directly changed to an action list.
    - what do we already have?
    - what has not been implemented yet?

Related issue: https://github.com/PaddlePaddle/Paddle/issues/6821

**Comment by kavyasrinet** [comment]
@lcy-seso Thanks so much for the writeup. This is very helpful. As discussed in the Hi group, let's pick the items one by one, discuss each of them and figure out a plan.

**Comment by lcy-seso** [comment]
The transformer also follows the encoder-decoder architecture. Encoder and decoder are stacks of many identical modules.

Computations/operators each part requires:
- Positional Embedding.
    - Nothing special, positional embedding is the sum of the positional encoding and the common word embeddings.

- Encoder is made of (1) self-attention part; (2)  Position-wise Feed-Forward network part;
    1. Multi-head self-attention
        - matrix multiplication
        - dot product
        - scale
        - sequence softmax
        - residul conection (addition)
        - layer normalization
    1. Position-wise Feed-Forward networks
        - two linear transformations (matrix multiplication and add, namely, the fully connected layer)
        - RELU activation
        - residual connection (addition)
        - layer normalization
- Decoder
    1. the operators used by the decoder are the same as that used by the encoder.

Because Transformer does not depend on any recurrence and convolutions, there are not many operators we need to implement. I think the only one is the layer normalization.

But,  one difficult I think is how to implement self-attention efficiently. Theoretically, this step can be implemented in high parallelization. I guess we need `while_op` or `dynamic RNN` for the first version.

The architecture of transformer is quite simple. I guess there will be many tricks to tune it once we can successfully run the model.

**Comment by lcy-seso** [comment]
I add a brief TODO list first for our discussion. The list needs priorities and can be done in parallel.

- Implement RNN encoder-decoder without attention.
    - prepare the English-Germany dataset.
    - add python wrapper for needed operators.
        - GRU, sequence softmax, GRU unit.
    - smoothly trained both on CPU/GPU / stably converge.
        - Currently, bugs occur in GPU mode.
- Implement RNN encoder-decoder with attention.
    - This model is not implemented and tested yet.
    - add all the needed python wrappers.
    - verify the attention which depends on dynamic RNN.

- Finish the standard beam search based on RNN encoder-decoder without attention.
    - Complete the code.
    - Give a basic example.

- Add the layer normalization operators.

- Build the transformer module by module.
    - Implement the self-attention. We can just debug self-attention (make it smoothly run) first on a simple task like text classification.
    - wrap the multi-head attention.
    - wrap the masked multi-head attention for the decoder.
    - wrap the positional embeddings.
    - wrap the basic computation block:
        - wrap the multi-head self-attention + residual connect + layer normalization.
        - wrap the position-wise feed-forward networks.
- Also, need to learn and try [Transformer in t2t](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py) to make clear of some detail tricks.
- The mechanism we make use of to implement self-attention (dynamic rnn / RNN block / while_op, exactly which one I am not sure right now) can also be used to implement attention in ConvS2S. Finally, I think we can also support the ConvS2S model.
- Verify beam search for the transformer.

**Comment by lcy-seso** [comment]
The following picture is from [googleblog](https://research.googleblog.com/2017/08/transformer-novel-neural-network.html), which well demonstrates how the transformer works.

<p align="center">
<img src="https://3.bp.blogspot.com/-aZ3zvPiCoXM/WaiKQO7KRnI/AAAAAAAAB_8/7a1CYjp40nUg4lKpW7covGZJQAySxlg8QCLcBGAs/s1600/transform20fps.gif" width=80%>
<br>How transformer works.
</p>

**Comment by pkuyym** [comment]
Currently, I have surveyed some implementation for Transformer. 
The authoritative version is: https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py
Besides, there is a simple and direct repo implemented by PyTorch: https://github.com/jadore801120/attention-is-all-you-need-pytorch

**Comment by lcy-seso** [comment]
We need still a better design for Transformer. `while_op` is important for many seq2seq tasks, but it is not necessary for Transformer because `while_op` still compute timestep by timestep.

@pkuyym  If you have already surveyed some implementation for the transformer, I  have one question, how they implement the self-attention both time and memory efficiently?  Does it need some special operator, such as "broadcast"? Is self-attention based on some elementary operators or we can just write a very specific operator for it?

**Comment by pkuyym** [comment]
@lcy-seso Yes, `while_op` is not necessary. 

I only surveyed the PyTorch version carefully to check the details of Transformer.

1. how they implement the self-attention both time and memory efficiently?

`torch.bmm` is the key operator in the implement of self-attention, I think parallelization between time steps is considered in the operator. I will paste some snippets to make things clear:

```
q_s = q.repeat(n_head, 1, 1).view(n_head, -1, d_model) # n_head x (mb_size*len_q) x d_model
k_s = k.repeat(n_head, 1, 1).view(n_head, -1, d_model) # n_head x (mb_size*len_k) x d_model
v_s = v.repeat(n_head, 1, 1).view(n_head, -1, d_model) # n_head x (mb_size*len_v) x d_model

# treat the result as a (n_head * mb_size) size batch
q_s = torch.bmm(q_s, self.w_qs).view(-1, len_q, d_k)   # (n_head*mb_size) x len_q x d_k
k_s = torch.bmm(k_s, self.w_ks).view(-1, len_k, d_k)   # (n_head*mb_size) x len_k x d_k
v_s = torch.bmm(v_s, self.w_vs).view(-1, len_v, d_v)   # (n_head*mb_size) x len_v x d_v

outputs, attns = self.attention(q_s, k_s, v_s, attn_mask=attn_mask.repeat(n_head, 1, 1))
```
The implementation of ```attention``` function is:

```
attn = torch.bmm(q, k.transpose(1, 2)) / self.temper
if attn_mask is not None:
    assert attn_mask.size() == attn.size(), \
        'Attention mask shape {} mismatch ' \
        'with Attention logit tensor shape ' \
        '{}.'.format(attn_mask.size(), attn.size())
    attn.data.masked_fill_(attn_mask, -float('inf'))

attn = self.softmax(attn)
attn = self.dropout(attn)
output = torch.bmm(attn, v)
```
As we can see, the key operation is ```torch.bmm``` which is short for ```batch mat mul```. The above implementation consumes much more memory to achieve parallelization between time steps.  

2. Does it need some special operator, such as "broadcast"?
Tile the key, value and query `head` times to exploit the efficient `bmm` operator.

3. Is self-attention based on some elementary operators or we can just write a very specific operator for it?
Since the above implementation take the input as padded tensor, I think we need do some adaption to support no-padding tensor.

**Comment by lcy-seso** [comment]
> how they implement the self-attention both time and memory efficiently?

I think I understand the point. Becuase both ConS2S and transformer use the "dot production attention". This kind of attention is very special. The process of "dot production" is equal to outer production which can be computed highly efficiently.

But the outer production cannot support variable length sequence. Whether we need padding, I will think more about this.

**Comment by pkuyym** [comment]
It seems that implementation based on `while_op` is not an optimal choice. Please think how to remove the padding part.

**Comment by lcy-seso** [comment]
The padding cannot be removed from transformer and ConvS2S.

**Comment by lcy-seso** [comment]
I close this issue first. We can discuss the problem found in a new issue or reopen this issue if needed.  Thanks, everyone.

**Comment by tigerneil** [comment]
http://nlp.seas.harvard.edu/2018/04/03/attention.html
